### PR TITLE
Cleanup TermsHashPerField

### PR DESCRIPTION
**Pull request by s1monw**
Several classes within the IndexWriter indexing chain haven't been touched for several years. Most of these classes expose their internals through public members and are difficult to construct in tests since they depend on many other classes. This change tries to clean up TermsHashPerField and adds a dedicated standalone test for it to make it more accessible for other developers since it's simpler to understand. There are also attempts to make documentation better as a result of this refactoring.

### CODE REVIEW

File: lucene/core/src/java/org/apache/lucene/index/TermsHashPerField.java [hunk 1]
```diff
@@ -19,203 +19,207 @@
...
+    // Init stream slices
+    // TODO: figure out why this is 2*streamCount here. streamCount should be enough?
+    if ((2*streamCount) + intPool.intUpto > IntBlockPool.INT_BLOCK_SIZE) {
+      // can we fit all the streams in the current buffer?
+      intPool.nextBuffer();
+    }
 
-      postingsArray.intStarts[termID] = intUptoStart + intPool.intOffset;
+    if (ByteBlockPool.BYTE_BLOCK_SIZE - bytePool.byteUpto < (2*streamCount) * ByteBlockPool.FIRST_LEVEL_SIZE) {
+      // can we fit at least one byte per stream in the current buffer, if not allocated a new one
```
Comment (mikemccand) [inline lucene/core/src/java/org/apache/lucene/index/TermsHashPerField.java:143]: s/`allocated`/`allocate`

File: lucene/core/src/java/org/apache/lucene/index/FreqProxTermsWriterPerField.java [hunk 2]
```diff
@@ -207,8 +202,6 @@ public void newPostingsArray() {
 
   @Override
   ParallelPostingsArray createPostingsArray(int size) {
-    IndexOptions indexOptions = fieldInfo.getIndexOptions();
-    assert indexOptions != IndexOptions.NONE;
```
[the remark this answers is not in the document]
Comment (s1monw) [reply lucene/core/src/java/org/apache/lucene/index/FreqProxTermsWriterPerField.java:211]: I moved it in a better place in the ctor of the base class. I think that's enough?

File: lucene/core/src/java/org/apache/lucene/index/FreqProxTermsWriterPerField.java [hunk 3]
```diff
@@ -56,12 +56,6 @@ public FreqProxTermsWriterPerField(FieldInvertState invertState, TermsHash terms
   @Override
   void finish() throws IOException {
     super.finish();
-    sumDocFreq += fieldState.uniqueTermCount;
-    sumTotalTermFreq += fieldState.length;
```
[the remark this answers is not in the document]
Comment (s1monw) [reply lucene/core/src/java/org/apache/lucene/index/FreqProxTermsWriterPerField.java:60]: `sumDocFreq` and `sumTotalTermFreq` are unused. They were used in `FreqProxFields` in the past but not anymore for a while now. I removed their commented out usage so you can see it in a followup commit

File: lucene/core/src/java/org/apache/lucene/index/TermVectorsConsumerPerField.java [hunk 4]
```diff
@@ -222,7 +234,7 @@ void writeProx(TermVectorsPostingsArray postings, int termID) {
   }
 
   @Override
-  void newTerm(final int termID) {
+  void newTerm(final int termID, final int docID) {
```
[the remark this answers is not in the document]
Comment (s1monw) [reply lucene/core/src/java/org/apache/lucene/index/TermVectorsConsumerPerField.java:237]: that's correct.

File: lucene/core/src/java/org/apache/lucene/index/TermsHashPerField.java [hunk 5]
```diff
@@ -19,203 +19,207 @@
...
+  TermsHashPerField(int streamCount, IntBlockPool intPool, ByteBlockPool bytePool, ByteBlockPool termBytePool,
+                    Counter bytesUsed, TermsHashPerField nextPerField, String fieldName, IndexOptions indexOptions) {
+    this.intPool = intPool;
+    this.bytePool = bytePool;
     this.streamCount = streamCount;
-    numPostingInt = 2*streamCount;
-    this.fieldInfo = fieldInfo;
+    this.fieldName = fieldName;
     this.nextPerField = nextPerField;
+    assert indexOptions != IndexOptions.NONE;
```
[the remark this answers is not in the document]
Comment (s1monw) [reply lucene/core/src/java/org/apache/lucene/index/TermsHashPerField.java:69]: yeah :)

**Comment by s1monw** [comment]
> I think we should run indexing throughput benchmark to make sure there's no real impact here. I'll do that.

+1 thanks @mikemccand

File: lucene/core/src/java/org/apache/lucene/index/FreqProxTermsWriterPerField.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (s1monw) [reply lucene/core/src/java/org/apache/lucene/index/FreqProxTermsWriterPerField.java:60]: see this https://github.com/apache/lucene-solr/pull/1573/files#diff-aa6c5376b6b755262430916164fd0088L84

File: lucene/core/src/java/org/apache/lucene/index/TermsHashPerField.java [hunk 6]
```diff
@@ -19,203 +19,207 @@
...
-
-  final int streamCount;
-  final int numPostingInt;
-
-  protected final FieldInfo fieldInfo;
-
-  final BytesRefHash bytesHash;
+  // for each term we store an integer per stream that points into the bytePool above
+  // the address is updated once data is written to the stream to point to the next free offset
+  // this the terms stream. The start address for the stream is stored in postingsArray.byteStarts[termId]
```
Comment (dweiss) [inline lucene/core/src/java/org/apache/lucene/index/TermsHashPerField.java:45]: this the?

File: lucene/core/src/java/org/apache/lucene/index/TermsHashPerField.java [hunk 7]
```diff
@@ -19,203 +19,207 @@
...
   }
 
-  int[] sortedTermIDs;
+  private int[] sortedTermIDs;
 
   /** Collapse the hash table and sort in-place; also sets
-   * this.sortedTermIDs to the results */
-  public int[] sortPostings() {
+   * this.sortedTermIDs to the results
+   * This method should not be called twice unless {@link #reset()}
```
Comment (dweiss) [inline lucene/core/src/java/org/apache/lucene/index/TermsHashPerField.java:97]: should not -> must not perhaps?

File: lucene/core/src/test/org/apache/lucene/index/TestTermsHashPerField.java [hunk 8]
```diff
@@ -0,0 +1,210 @@
...
+          postings.lastDocIDs[termID] = docID;
+        } else {
+          postings.termFreqs[termID] = Math.addExact(postings.termFreqs[termID], 1);
+        }
+      }
+
+      @Override
+      void newPostingsArray() {
+        freqProxPostingsArray = (FreqProxTermsWriterPerField.FreqProxPostingsArray) postingsArray;
+
```
Comment (dweiss) [inline lucene/core/src/test/org/apache/lucene/index/TestTermsHashPerField.java:81]: extra space?

**Comment by mikemccand** [comment]
I tested indexing throughput on `luceneutil` with `wikimediumall`, single thread for indexing `SerialMergeScheduler`:

```
[mike@beast3 facet]$ grep "GB/hour" /l/logs/simon?
/l/logs/simon0:Indexer: 46.44432470391602 GB/hour plain text
/l/logs/simon1:Indexer: 46.267723012921515 GB/hour plain text
/l/logs/simon2:Indexer: 46.26414201429784 GB/hour plain text
[mike@beast3 facet]$ grep "GB/hour" /l/logs/trunk?
/l/logs/trunk0:Indexer: 45.632881600179495 GB/hour plain text
/l/logs/trunk1:Indexer: 46.09383252131896 GB/hour plain text
/l/logs/trunk2:Indexer: 45.439666582156924 GB/hour plain text
```

Net/net this change might be a bit faster, or just noise, so all clear to push!  Thanks @s1monw.

**Comment by s1monw** [comment]
thanks @mikemccand - I will run tests again and push.

**Comment by mikemccand** [comment]
> thanks @mikemccand - I will run tests again and push.

++

**Pull request merged**